pkg/alertmanager/rate_limited_notifier.go

```diff
@@ -0,0 +1,65 @@
+package alertmanager
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/prometheus/alertmanager/notify"
+	"github.com/prometheus/alertmanager/types"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.uber.org/atomic"
+	"golang.org/x/time/rate"
+)
+
+type rateLimits interface {
+	RateLimit() rate.Limit
+	Burst() int
+}
+
+type rateLimitedNotifier struct {
+	upstream notify.Notifier
+	counter  prometheus.Counter
+
+	limiter *rate.Limiter
+	limits  rateLimits
+
+	recheckInterval time.Duration
+	recheckAt       atomic.Int64 // unix nanoseconds timestamp
+}
+
+func newRateLimitedNotifier(upstream notify.Notifier, limits rateLimits, recheckInterval time.Duration, counter prometheus.Counter) *rateLimitedNotifier {
+	return &rateLimitedNotifier{
+		upstream:        upstream,
+		counter:         counter,
+		limits:          limits,
+		limiter:         rate.NewLimiter(limits.RateLimit(), limits.Burst()),
+		recheckInterval: recheckInterval,
+	}
+}
+
+var errRateLimited = errors.New("failed to notify due to rate limits")
+
+func (r *rateLimitedNotifier) Notify(ctx context.Context, alerts ...*types.Alert) (bool, error) {
+	now := time.Now()
+	if now.UnixNano() >= r.recheckAt.Load() {
+		if limit := r.limits.RateLimit(); r.limiter.Limit() != limit {
+			r.limiter.SetLimitAt(now, limit)
+		}
+
+		if burst := r.limits.Burst(); r.limiter.Burst() != burst {
+			r.limiter.SetBurstAt(now, burst)
+		}
+
+		r.recheckAt.Store(now.UnixNano() + r.recheckInterval.Nanoseconds())
+	}
+
+	// This counts as single notification, no matter how many alerts there are in it.
+	if !r.limiter.AllowN(now, 1) {
+		r.counter.Inc()
+		// Don't retry this notification later.
+		return false, errRateLimited
+	}
+
+	return r.upstream.Notify(ctx, alerts...)
```

Comment:
I think we should cancel the limiter reservation if the upstream returns error. WDYT?

Reply:
Hmm, I'm not sure. These limits will be already quite high. If user is hitting them, it's likely that user is doing something wrong or bad.

Imagine someone abusing alertmanager notifications for sending lot of requests to a website, which starts crashing and return 500. This can look like failed notification – if we cancel the reservation, it allows bad actor to keep sending more requests. (This PR is only dealing with emails, but we will reuse this for other integrations).

WDYT?

Reply:
I see your point and your example makes sense. I'm also thinking about the opposite case: a legit receiver backend server is down, retries hit the rate limit even if no notification has been successfully delivered. However, since we can't distinguish it, it's probably safer to keep the current logic as you suggest.

Reply:
I agree with @pstibrany we should count against the limit whether success or fail, because opens some possibilities for abuse/broken things to make it through if done the other way.